python.d/ceph.chart.py

```diff
@@ -0,0 +1,164 @@
+# -*- coding: utf-8 -*-
+# Description: ceph netdata python.d module
+# Author: Luis Eduardo (lets00)
+
+try:
+    import rados
+    CEPH = True
+except ImportError:
+    CEPH = False
+
+import json
+from bases.FrameworkServices.SimpleService import SimpleService
+
+# default module values (can be overridden per job in `config`)
+update_every = 10
+priority = 60000
+retries = 60
+
+ORDER = ['general', 'pool_usage', 'pool_objects', 'osd_usage']
+
+CHARTS = {
+    'general': {
+        'options': [None, 'Ceph General Options', 'KB', 'general', 'ceph.general', 'stacked'],
+        'lines': [
+            ['general_available', 'avail', 'absolute', 1, 1024],
+            ['general_usage', 'used', 'absolute', 1, 1024]
+        ]
+    },
+    'pool_usage': {
+        'options': [None, 'Ceph Pools', 'KB', 'pool', 'ceph.pool_usage', 'line'],
+        'lines': []
+    },
+    'pool_objects': {
+        'options': [None, 'Ceph Pools', 'objects', 'pool', 'ceph.pool_objects', 'line'],
+        'lines': []
+    },
+    'osd_usage': {
+        'options': [None, 'Ceph OSDs', 'KB', 'osd', 'ceph.osd_usage', 'line'],
+        'lines': []
+    }
+}
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+        self.config_file = self.configuration.get('config_file')
+        self.keyring_file = self.configuration.get('keyring_file')
+        self.cluster = rados.Rados(conffile=self.config_file,
+                                   conf=dict(keyring=self.keyring_file))
```

Comment:
Сan this cause an exception?
I mean if there is no `self.config_file` or `self.keyring_file` or `chmod 640 /etc/ceph/ceph.client.admin.keyring` not done.

Reply:
that is what i got
```python
>>> import rados
>>> cluster = rados.Rados(conffile=None, conf=dict(keyring=None))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "rados.pyx", line 525, in rados.Rados.__init__ (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:6445)
  File "rados.pyx", line 425, in rados.requires.wrapper.validate_func (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:4657)
  File "rados.pyx", line 571, in rados.Rados.__setup (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:7612)
  File "rados.pyx", line 424, in rados.requires.wrapper.validate_func (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:4634)
  File "rados.pyx", line 412, in rados.requires.check_type (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:4010)
TypeError: val must be basestring
>>> cluster = rados.Rados(conffile=str(), conf=dict(keyring=str()))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "rados.pyx", line 525, in rados.Rados.__init__ (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:6445)
  File "rados.pyx", line 425, in rados.requires.wrapper.validate_func (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:4657)
  File "rados.pyx", line 568, in rados.Rados.__setup (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:7414)
  File "rados.pyx", line 425, in rados.requires.wrapper.validate_func (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:4657)
  File "rados.pyx", line 631, in rados.Rados.conf_read_file (/build/ceph-HiUCwo/ceph-10.2.5/src/build/rados.c:8591)
rados.Error: error calling conf_read_file: error code 22
```

Reply:
And
```python
>>> cluster = rados.Rados(conffile='/etc/network/interfaces', conf=dict(keyring=str()))
>>> 
```
😄

Reply:
Oh, I got confused. Sorry about that =(. I tested this on passed, but not remember that this could be occour. 
In fact, the exception is only raised on self.cluster.connect() when the file (/etc/ceph/keyring...) does not correct permission. If conffile or keyring is None, an exception is raised before (when you create the cluster object). I will fix this.